Timer plugin

lsowa opened the pull request:
The Timer plugin enforces a recurring, time-of-day schedule for the demand of a target pool (see doc).

Implemented:

- [x] Timer module
- [x] pytests
- [x] documentation
- [x] tested in seven day long test set up

lsowa committed:
Implement Stopper plugin for demand management based on time

lsowa committed:
finished draft for timer plugin

lsowa committed:
Rename Stopper class to Timer and update references in configuration

lsowa committed:
Refactor str_to_time function for improved error handling and clarity

lsowa committed:
adapt latest_timestamp for easier testing

lsowa committed:
firts impl. of pytests

lsowa committed:
add documenantation and index

lsowa committed:
cleanup

lsowa committed:
correct documentation

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+        hours, minutes = int(hours_str), int(minutes_str)
+    except ValueError as exc:
+        raise ValueError(f"invalid time specification {value}")
+    return datetime_time(hour=hours, minute=minutes)
```
Unless I am missing some extra feature here, please use `datetime`'s native parsing. The equivalent of this function is `datetime.strptime(value, '%H:%M').time()`.

lsowa replied:
Replaced this function with `datetime.strptime(value, '%H:%M').time()`.

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+        super().__init__(target)
+
+        assert interval > 0, "Interval must be a positive integer."
+        self.interval = interval
```
Since you are working on a schedule, there is no need to have a polling interval. You can simply `asyncio.sleep` until the next schedule time.

I recommend to just remove it. See comments on the `run` method for auxiliary changes.

```suggestion
```

lsowa replied:
Dropped `interval` und adapted `run()` function (see other comment).

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+
+        schedule = {str_to_time(key): value for key,value in schedule.items()}
+        self.schedule = schedule
+        self.latest_sched_demand = self._refresh_demand()
```
This attribute doesn't seem to be used outside of a redundant setter. I recommend to just remove it.

```suggestion
```

lsowa replied:
Removed this as sugegsted.

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+        self,
+        target: Pool,
+        schedule: Schedule,
+        interval: int = 300,
```
See first comment on `self.interval`.
```suggestion
```

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+    @demand.setter
+    def demand(self, value: float) -> None:
+        # Ignore user supplied demand and always enforce the scheduled value.
+        self.target.demand = self.latest_sched_demand
```
`self.target.demand` should already equal `self.latest_sched_demand`. You don't have to do anything here.

```suggestion
        pass
```

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+        """Update the demand periodically according to the schedule."""
+        while True:
+            self._refresh_demand()
+            await asyncio.sleep(self.interval)
```
Since you are operating on a fixed schedule, there is no need to poll and fetch the current time slot again and again. Instead, endlessly [`cycle`](https://docs.python.org/3/library/itertools.html#itertools.cycle) through the schedule and sleep precisely until you have to take action.

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+
+
+@service(flavour=asyncio)
+class Timer(PoolDecorator):
```
Maybe this should be a `Controller` instead. Since it doesn't allow setting the `demand`, there seems to be no point adding further Decorators or Controllers around it.

lsowa replied:
Implemented the plugin as a `Controller`.

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+        self.interval = interval
+
+        schedule = {str_to_time(key): value for key,value in schedule.items()}
+        self.schedule = schedule
```
In `_refresh_demand` is an assertion that the `value` here is positive. Please do the validation here already. Since the data must be sorted for various algorithms, either checking or enforcing this here seems prudent.  
I also recommend to rename the `key`/`value` to what they actually represent.

```suggestion
        schedule = {str_to_time(time): demand for time, demand in schedule.items()}
        assert all(demand >= 0 for demand in schedule.values())
        self.schedule = dict(sorted(schedule.items()))
```

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+        assert self.latest_sched_demand >= 0.0
+        assert self.target.demand >= 0.0
+        
+        return self.latest_sched_demand
```
Since you are operating on a fixed schedule, there is no need to poll and fetch the current time slot again and again. Instead, I recommend to endlessly loop through the schedule and sleep precisely until you have to take action.
```suggestion
        # repeatedly sleep until the next scheduled change of demand
        # this automatically fast-forwards through past events, meaning
        # no special handling for the first day is needed
        today = date.today()
        while True:
            for start_time, demand in schedule.items():
                start_delta = datetime.combine(today, start_time) - datetime.now()
                await asyncio.sleep(start_delta.total_seconds())
                self.target.demand = demand
            today += timedelta(days=1)
```
The `timedelta` must also be imported from `datetime`.

lsowa replied:
I implemented this machenism by looping manually with an index `idx` over the scheduled times. Here it is straight forward to initialize an demand, which is relevant when `Timer` is started before the first time stamp.
I had to use trio, which seems to accept positive time deltas only, so I capped these to zero.

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+from cobald.interfaces import Pool, PoolDecorator
+
+TimeInput = Union[str, datetime_time]
+Schedule = Mapping[TimeInput, float]
```
The initializer always calls `str_to_time` on the `Schedule` key, meaning that only `str` keys are acceptable.

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,92 @@
+            return candidate
+    # All configured times are in the future relative to ``current_time``.
+    # return latest timestamp
+    return ordered_times[-1]
```
This is probably no longer needed, see later comments. If you want to keep it:
- `current_timestamp` or `recent_timestamp` is probably a better name (without knowing the implementation, "latest" is likely to be read as "most late" instead of "most recent").
-  ordering should already be a prerequisite and established outside this function. Sorting the input again and again is a waste since a) the actual user input is likely sorted anyways and b) it's simple to sort once when reading the input.

lsowa replied:
Dropped this since it's not needed after adapting the `run()` method.

maxfischer2781 reviewed (changes requested):
Thanks for the PR, this looks like a useful tool.

There are quite a lot of comments from me on the implementation since it can be done much more efficiently. Please take them into account unless you have a good reason not to.  
That said, I want to point out that your code was pretty good to understand; all these comments are there because you made it clear what you want to do.

Let's look at the docs and tests once the code has settled.

lsowa committed:
technical adaptions for timer

lsowa committed:
add doc strings & optimize naming

lsowa committed:
remove type alias for schedule parameter

lsowa committed:
rename keys to sched_times for clarity

lsowa committed:
better variable names in schedule mapping

lsowa commented:
Thank you very much for your useful comments! I implemented all of them and the code is now better streamlined. I also tested it again.

Overview:
- Implemented plugin as `Controler`
- Used `datetime` to drop complicated custom functions
- Implemented proposed sleep method to schedule demand.

I'll adapt documentation and tests when we have a fixed version of the code.

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,59 @@
+from datetime import date, datetime, timedelta
+from typing import Mapping
+
+import trio
```
Is there a reason why you switched to `trio`? You need only one function and the one from `asyncio` is slightly better for your usecase (it automatically caps negative values).

Generally, `asyncio` has come a long way since COBalD was made initially. It should be your preferred event loop unless there are strong technical reasons to pick something else.

lsowa replied:
I oriented myself on the other controllers, which all use `trio`, but I accidentally mixed `@service(flavour=trio)` with `asyncio.sleep`. I now switched fully to asyncio.

maxfischer2781 commented on cobald_hep_plugins/timer.py:
```diff
@@ -0,0 +1,59 @@
+
+        start_time = datetime.now().time()
+        idx = bisect.bisect_right(sched_times, start_time) - 1
+        idx %= len(sched_times)
```
I strongly recommend not to use explicit indexing. It currently serves tripple duty for initial, stepwise and daily increment, with a magic number attached to each of that.

Since you are capping negative sleep delays (either manually or via `asyncio.sleep`) you don't need the initial increment. You already rely on the loop fast-forwarding through past events, and it's always possible (even if unlikely) that this also applies during the loop.

That leaves only the stepwise and daily increment, which should exactly match a regular `for` loop over `self.schedule` and a trailing increment, respectively.

lsowa replied:
Thanks for the clarification. I agree that the explicit indexing was doing too much here. I have updated the implementation to follow your proposed approach.

maxfischer2781 reviewed (changes requested):
Sorry, got some comments on the implementation still.

lsowa committed:
move from trio to asyncio

lsowa committed:
adapt run() funcrion

lsowa commented:
Thanks again for your comments! When we agree on the final code, I will adapt the tests and documentation accordingly.

maxfischer2781 reviewed (commented):
Code looks fine by me, let's continue with the docs and such.

lsowa committed:
doc

lsowa committed:
update tests

lsowa committed:
misc

lsowa committed:
actually use instances of MockPool, misc

lsowa committed:
Ignore flake8 B905 when using `zip()` without `strict` arg to be compatible with Python < 3.10

lsowa committed:
Apply global black formatting to pass tests (not part of this PR but since no parallel work is ongoing it's ok I guess)

lsowa commented:
Thank you! I updated the docs and tests accordingly.

This is not really subject of this PR, but regarding the pipelines:
- I applied the black formatter on a few more files to allow the static checks to run through [b8b5b4904dc6d6ab2f51ea888717735a00568b54](https://github.com/MatterMiners/cobald-hep-plugins/commit/b8b5b4904dc6d6ab2f51ea888717735a00568b54)
- There still seems to be an issue with the builds for different python versions: we're testing `python3.6` and `python3.7` on `ubuntu-latest` which seems not to be supported. I can fix that by dropping the old/not compatible python versions or test them with a fixed `ubunut-XX` runner. Either in an seperate or in this PR.

giffels commented:
>  There still seems to be an issue with the builds for different python versions: we're testing `python3.6` and `python3.7` on `ubuntu-latest` which seems not to be supported. I can fix that by dropping the old/not compatible python versions or test them with a fixed `ubunut-XX` runner. Either in an seperate or in this PR.

SHould be fixed in #7

lsowa committed:
Revert "Apply global black formatting to pass tests (not part of this PR but since no parallel work is ongoing it's ok I guess)"

Revert changes made by black formatter - this was already taken care of
in another PR. This reverts commit b8b5b4904dc6d6ab2f51ea888717735a00568b54.

lsowa commented:
> > There still seems to be an issue with the builds for different python versions: we're testing `python3.6` and `python3.7` on `ubuntu-latest` which seems not to be supported. I can fix that by dropping the old/not compatible python versions or test them with a fixed `ubunut-XX` runner. Either in an seperate or in this PR.
> 
> SHould be fixed in #7

Thank you! As soon as #7 is merged I'll update my fork.

lsowa committed:
Merge branch 'MatterMiners:master' into mirror_orig_repo

lsowa commented:
I found this old PR. Only the doc review was open here. I just re-checked it and it looks good to me.
